src/jit/assertionprop.cpp

```diff
@@ -4446,47 +4469,69 @@ ASSERT_TP* Compiler::optComputeAssertionGen()
 
             for (GenTreePtr tree = stmt->gtStmt.gtStmtList; tree; tree = tree->gtNext)
             {
-                // Store whatever we have accumulated into jumpDest edge's valueGen.
                 if (tree->gtOper == GT_JTRUE)
                 {
-                    BitVecOps::Assign(apTraits, jumpDestValueGen, valueGen);
+                    assert((tree->gtNext == nullptr) && (stmt->gtNext == nullptr));
+                    jtrue = tree;
+                    break;
                 }
-                if (!tree->HasAssertion())
+
+                AssertionIndex index = tree->GetAssertion();
+                if (index != NO_ASSERTION_INDEX)
                 {
-                    continue;
+                    optImpliedAssertions(index, valueGen);
+                    BitVecOps::AddElemD(apTraits, valueGen, index - 1);
                 }
+            }
+        }
 
-                // For regular trees, just update valueGen. For GT_JTRUE, for false part,
-                // update valueGen and true part update jumpDestValueGen.
-                AssertionIndex assertionIndex[2] = {(AssertionIndex)tree->GetAssertion(),
-                                                    (tree->OperGet() == GT_JTRUE)
-                                                        ? optFindComplementary((AssertionIndex)tree->GetAssertion())
-                                                        : 0};
+        if (jtrue != nullptr)
+        {
+            // Copy whatever we have accumulated into jumpDest edge's valueGen.
+            ASSERT_TP jumpDestValueGen = BitVecOps::MakeCopy(apTraits, valueGen);
 
-                for (unsigned i = 0; i < 2; ++i)
+            AssertionIndex index = jtrue->GetAssertion();
```

Comment:
Wouldn't mind a comment here mentioning that the index stored on a `GT_JTRUE` is `index | optAssertionEdge`.  (Actually if I had my druthers there'd be a layer of abstraction here like separate accessors for a node's assertion index and its edge kind that do the masking [or perhaps the plain `GetAssertion()` accessor would assert that it's not called on a jtrue and that the edge bits aren't set, rather than masking, and callers would use a different `GetJTrueAssertion()` to get the masking], but I don't feel too strongly about it)